Step3 Doodle 이미지를 보여주는 화면 구현

seizze opened the pull request:
### 구현 내용
* Doodle 이미지를 보여주는 화면까지 구현하였습니다.

### 고민한 내용
* 컬렉션뷰 컨트롤러를 생성하는 코드를 간소화하기 위해 convenience init을 추가하였습니다.
* DoodleViewController는 처음엔 셀 개수를 0으로 하여 컬렉션뷰를 그리고, 원격에서 JSON 데이터를 받아오는 일이 완료되었다는 알림을 받아 다시 컬렉션뷰를 로드하도록 구현하였습니다.
* 비동기적으로 원격 이미지를 가져올 때 글로벌 큐에 Task를 전달하였고, 이때 QoS를 Utility로 지정하였습니다. 유저에게 보이는 네트워크 통신등에는 utility가 적합하다고 하여 이렇게 하였습니다. 또한 UI를 업데이트하는 Task는 메인 큐에 dispatch하도록 하였습니다.
* 원격에서 데이터를 가져오는 객체는 Data Source와만 연결되도록 설계하였습니다.
* DoodleViewController를 dismiss할 때 메모리에서 해제되는지 deinit에 브레이크 포인트를 찍어서 확인하였습니다.

### 배운 내용
* 컬렉션뷰 컨트롤러 생성시에 UICollectionViewLayout()가 아닌 UICollectionViewFlowLayout()을 전달해야 한다는 것을 배웠습니다.

seizze committed:
feat: 스토리보드에서 navigation controller 삭제

seizze committed:
feat: Navigation controller를 코드로 생성

seizze committed:
feat: Superview를 채우는 uiview extension 추가

seizze committed:
feat: Collectionview cell을 코드로 생성

seizze committed:
fix: 컬렉션뷰 안뜨는 오류 수정

seizze committed:
feat: Doodle VC 추가

seizze committed:
feat: 왼쪽 bar button item 추가

seizze committed:
feat: 화면 전환 구현

Limwin94 committed:
Merge pull request #26 from codesquad-memeber-2020/feature/display-doodle_issue-025

뷰 구조 변경

Limwin94 committed:
Merge pull request #28 from codesquad-memeber-2020/feature/display-doodle_issue-027

Doodle View Controller 화면 셋업

Limwin94 committed:
json decode 객체 생성

Limwin94 committed:
resolve merge confilct

seizze committed:
Merge pull request #29 from codesquad-memeber-2020/feature/display-doodle_issue-024

decoder 기능 추가.

seizze committed:
feat: GCD 이용 이미지 디스플레이

seizze committed:
Merge pull request #31 from codesquad-memeber-2020/feature/display-doodle_issue-030

GCD를 이용하여 이미지 디스플레이

seizze committed:
docs: Update README.md

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,75 @@
+    }
+    
+    convenience init() {
+        self.init(collectionViewLayout: UICollectionViewFlowLayout())
```
뷰 컨트롤러는 convenience init() 만드는 게 위험합니다. 
차라리 의존성을 주입하는 setter를 만들걸나 프로토콜을 채택하는 게 좋습니다.

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,75 @@
+
+extension DoodleViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        let size = collectionView.frame.width / 3 - 1
```
사이즈를 매번 계산하지 않아도 되는거면 화면이 보일 때 계산하는 게 좋습니다.

godrm commented on PhotosApp/PhotosApp/Model/DoodleDataSource.swift:
```diff
@@ -0,0 +1,31 @@
+        }
+        
+        DispatchQueue.global(qos: .utility).async { [weak self] in
+            let image = self?.remoteSource.fetchImage(index: indexPath.item)
```
셀을 표시할 때 바로 fetch하기 보다는 미리 다운로드 받을 수 있는 prefetch 방식을 사용하는 게 좋습니다.

godrm commented on PhotosApp/PhotosApp/Model/DoodleDataSource.swift:
```diff
@@ -0,0 +1,31 @@
+        
+        DispatchQueue.global(qos: .utility).async { [weak self] in
+            let image = self?.remoteSource.fetchImage(index: indexPath.item)
+            DispatchQueue.main.async { cell.setDoodle(to: image) }
```
다만 이렇게 하면 다운로드도 비동기라서 셀을 스크롤하다보면 다른 이미지가 보여지는 경우가 생길 수 있습니다. 그 부분을 반드시 주의해야 합니다.

godrm commented on PhotosApp/PhotosApp/Model/JsonDecoder.swift:
```diff
@@ -0,0 +1,48 @@
+import Foundation
+import UIKit
+
+class JsonDecoder {
```
이름은 하는 기능에 비해서 일반적인 이름이네요.

godrm commented on PhotosApp/PhotosApp/Model/JsonDecoder.swift:
```diff
@@ -0,0 +1,48 @@
+                let doodleImgDataList = try? decoder.decode([DoodleImageData].self, from: data) {
+                self.doodleImgData = doodleImgDataList
+                DispatchQueue.main.async {
+                    NotificationCenter.default.post(name: .doodleDidChange, object: nil)
```
노티를 보내는 건 GCD를 사용하지 않아도 됩니다.

godrm commented on PhotosApp/PhotosApp/Model/JsonDecoder.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+    
+    func fetchImage(index: Int) -> UIImage {
+        guard let data = try? Data(contentsOf: doodleImgData[index].image),
```
다운로드는 비동기 방식이 아니라 동기 방식이네요. 이 부분도 비동기 방식으로 변경해보세요.

seizze replied:
질문이 있습니다.🧐
이 fetchImage 메서드의 경우 collectionView(_:cellForItemAt:)에서 글로벌 큐에 async로 전달하여 수행하게 했는데, 그러면 상위 함수에서 비동기 방식이 되는거 아닌가요? fetchImage도 비동기로 변경해야 비동기가 되는것인가요??

godrm replied:
그건 아니죠. 상위에서 동기인지 비동기인지 알 수 있으면 다행인데 그렇지 않으면 내부에서 비동기로 처리하는게 좋지 않을까요

seizze replied:
음....... 제가 이해를 잘 한 건지 모르겠습니다.
오래 걸리는 작업을 상위에서 직접 비동기 큐에 넣어서 사용하는 것보다는 상위 함수에서는 호출만 하고 내부에서 자체적으로 비동기적으로 동작하는 편이 좋다고 말씀하신거 맞나요?? ㅎㅎ

godrm replied:
그걸 명확하게 하는 게 좋은거죠. 매번 호출할 때마다 상위에서 비동기로 처리하기 보다 어차피 비동기로 처리해야 한다면 하위 클래스 동작을 비동기로 만드는 게 낫겠죠. 
URLSession에 DataTask 같은 객체들 처럼 네트워크에 요청하는 객체는 비동기라는 것을 명시하는 게 좋습니다.

seizze replied:
아하 네 알겠습니다! 고쳐보겠습니다.

godrm commented on PhotosApp/PhotosApp/PhotoViewController.swift:
```diff
@@ -0,0 +1,94 @@
+    
+    @objc private func showDoodles() {
+        let nextViewController = UINavigationController(rootViewController: DoodleViewController())
+        present(nextViewController, animated: true)
```
이 부분이 이상합니다. 다음 화면만 내비게이션을 시키는 게 아니라 ViewController로 부터 내비게이션 컨트롤러에 포함되어 있어야 합니다.

godrm commented on PhotosApp/PhotosApp/PhotoViewController.swift:
```diff
@@ -0,0 +1,94 @@
+            })
+            if let changed = changes[.changed] as? [IndexPath], !changed.isEmpty {
+                collectionView.reloadItems(at: changed)
+            }
```
여기 비교문이 복합하네요. 이렇게 할 수 밖에 없는걸까요?

godrm commented on PhotosApp/PhotosApp/SceneDelegate.swift:
```diff
@@ -11,43 +11,16 @@ import UIKit
+            let rootViewController = PhotoViewController()
+            window.rootViewController = UINavigationController(rootViewController: rootViewController)
+            self.window = window
+            window.makeKeyAndVisible()
```
아 여기 이미 뷰컨트롤러에 내비게이션 컨트롤러를 포함했군요! 
그러면 위에 DoobleViewController를 표시할 때는 self.navigationContoller.pushXXX 이렇게 처리해도 됩니다.

seizze replied:
앗 이것은 미션 내용에서 네비바에 Doodles라고 쓰여진 화면이 모달로 올라온 것처럼 보여서 화면 전환 시에 새 내비게이션 컨트롤러를 만들고 그 안에 DoodleViewController를 임베드하여 모달로 띄웠었습니다!

DoodleViewController만 생성해서 push하는 방식으로 하면 옆으로 넘어가는 화면 전환이 일어나는거 아닌가요? push를 해도 아래에서 올라오는 방법이 있는건가요? 🤔

godrm replied:
아뇨. 내비게이션이 있어야 push가 가능하고 push하면 우측에서 나타나고,
아래서 올라오는 방식은 modal로 present를 하는 겁니다. 
modal로 할때는 내비게이션 컨트롤러가 필요없다는 겁니다.

seizze replied:
넵 이것도 고쳐보겠습니다!

godrm commented on PhotosApp/PhotosApp/Util/UIView+FillSuperView.swift:
```diff
@@ -0,0 +1,19 @@
+        topAnchor.constraint(equalTo: superview.topAnchor).isActive = true
+        leadingAnchor.constraint(equalTo: superview.leadingAnchor).isActive = true
+        trailingAnchor.constraint(equalTo: superview.trailingAnchor).isActive = true
+        bottomAnchor.constraint(equalTo: superview.bottomAnchor).isActive = true
```
이게 편리하긴 하지만 조심해야 합니다. 이미 제약사항이 있는 경우가 있을 수도 있고
UIView 전체를 확장한 것이기 때문에 다른 뷰에도 영향을 줄 수 있습니다. 참고하세요.

seizze replied:
넵 조심해서 사용하겠습니다! 🙏

godrm commented on PhotosApp/PhotosApp/View/DoodleCell.swift:
```diff
@@ -0,0 +1,44 @@
+    }
+    
+    convenience init() {
+        self.init(frame: .zero)
```
뷰 객체에 convenience init()을 만드는 건 나만 쓰는 메소드가 되서, 다른 메소드로 생성하면 제대로 동작하지 않을 수 있습니다. 어떤 생성자로 생성해도 동일한 동작을 하도록 만드는 게 좋습니다.

godrm commented on PhotosApp/PhotosApp/View/DoodleCell.swift:
```diff
@@ -0,0 +1,44 @@
+    }
+    
+    func setDoodle(to image: UIImage?) {
+        guard let image = image else { return }
```
이렇게 할꺼면 왜 옵셔널로 했나요?

godrm commented on PhotosApp/PhotosApp/View/PhotoCell.swift:
```diff
@@ -9,10 +9,35 @@
+    private func setupView() {
+        translatesAutoresizingMaskIntoConstraints = false
+        addSubview(photoImageView)
+        photoImageView.fillSuperView()
```
위에 DoodleCell 이랑 구현내용이 동일한거 아닌가요?

godrm reviewed (changes requested):
다음 내용을 확인해보세요.

Limwin94 committed:
[#32] feat: Add UIGestureRecognizer

Doodle image long press gesture action add

seizze committed:
[#34] refactor: Re-arrange project file tree

Limwin94 committed:
[#32] feat: Set select menu to float DoodleView

seizze committed:
[#34] refactor: Remove convenience initializer

Also remove other initializers from all collection view controllers

Limwin94 committed:
[#32] feat: save cell image

if click 'save' menu, download the image to default photo album.

Limwin94 committed:
[#32] style: set photo content mode

Limwin94 committed:
[#32] refactor: add guard let context in saveImage metho

seizze committed:
Merge pull request #35 from codesquad-memeber-2020/feature/display-doodle_issue-032

[#32] long press gesture 구현

seizze committed:
[#34] refactor: Rename some types and variables

Rename JsonDecoder to DoodleAPIClient
Rename some variables related to networking

seizze committed:
[#34] refactor: Improve cell size calculating

seizze committed:
[#34] merge: Resolve merge conflict

seizze committed:
Merge pull request #37 from codesquad-memeber-2020/feature/display-doodle_issue-034

클래스들 간 관계 개선

Limwin94 commented:
@seizze 이거 왜 머지 안했었지...?

godrm commented:
으응? 언제리뷰한거죠 ㅋㅋ
